src/Hazelcast.Net/CP/SessionAwareSemaphore.cs

```diff
@@ -0,0 +1,230 @@
+            {
+                _sessionManager.ReleaseSession(CPGroupId, sessionId, permits);
+                throw;
+            }
+        }
+    }
+
+    public async Task<bool> TryAcquireAsync(int permits = 1, long timeoutMs = 0)
+    {
+        permits.ThrowIfLessThanOrZero(nameof(permits));
+        if (timeoutMs < 0) timeoutMs = 0;
+        var invocationUid = Guid.NewGuid();
+        var countdown = new Countdown(timeoutMs);
+        var threadId = GetThreadId();
+
+        for (;;)
+        {
+            var sessionId = await _sessionManager.AcquireSessionAsync(CPGroupId, permits).CfAwait();
+
+            try
+            {
+                var requestMessage = SemaphoreAcquireCodec.EncodeRequest(
+                    CPGroupId, Name, sessionId, threadId, invocationUid, permits, 
+                    countdown.RemainingMilliseconds);
+                var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+                var acquired = SemaphoreAcquireCodec.DecodeResponse(responseMessage).Response;
+                if (!acquired) _sessionManager.ReleaseSession(CPGroupId, sessionId, permits);
+                return acquired;
+            }
+            catch (RemoteException e) when (e.Error == RemoteError.SessionExpiredException)
+            {
+                _sessionManager.InvalidateSession(CPGroupId, sessionId);
+                if (countdown.Elapsed) return false; // else try again
+            }
+            catch (RemoteException e) when (e.Error == RemoteError.WaitKeyCancelledException)
+            {
+                _sessionManager.ReleaseSession(CPGroupId, sessionId, permits);
+                return false;
+            }
+            catch (Exception)
+            {
+                _sessionManager.ReleaseSession(CPGroupId, sessionId, permits);
+                throw;
+            }
+        }
+    }
+
+    public async Task ReleaseAsync(int permits = 1)
+    {
+        permits.ThrowIfLessThanOrZero(nameof(permits));
+        var invocationUid = Guid.NewGuid();
+        var threadId = GetThreadId();
+
+        var sessionId = _sessionManager.GetSessionId(CPGroupId);
+        if (sessionId == CPSessionManager.NoSessionId)
+            throw new InvalidOperationException("No CP session.");
+
+        try
+        {
+            var requestMessage = SemaphoreReleaseCodec.EncodeRequest(CPGroupId, Name, sessionId, threadId, invocationUid, permits);
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            _ = SemaphoreReleaseCodec.DecodeResponse(responseMessage);
+        }
+        catch (RemoteException e) when (e.Error == RemoteError.SessionExpiredException)
+        {
+            _sessionManager.InvalidateSession(CPGroupId, sessionId);
+            throw new InvalidOperationException(
+                $"Could not release semaphore {Name} because the release operation was cancelled, possibly because of another operation.");
```

Comment:
I'm confused with exception message. The error is about expiration but message says cancelled.

Reply:
indeed, should be different. Stephan to verify all exception messages.

Reply:
Fixed in f41fb4088